src/main/java/com/google/step/datamanager/DealManagerDatastore.java

```diff
@@ -99,54 +139,168 @@ public Deal updateDeal(Deal deal) {
+    List<Deal> dealResults = new ArrayList<>();
+    List<Long> idsOfFollowedFieldName = getFollowedSomething(userId, fieldName);
+    for (Long id : idsOfFollowedFieldName) {
+      Filter propertyFilter = new FilterPredicate(filterAttribute, FilterOperator.EQUAL, id);
+      Query query = new Query("Deal").setFilter(propertyFilter);
+      PreparedQuery pq = datastore.prepare(query);
+      for (Entity entity : pq.asIterable()) {
+        dealResults.add(readDeal(entity.getKey().getId()));
+      }
+    }
+    return dealResults;
+  }
+
+  /**
+   * This is copied from the follow manager method which is private, would it be better to make the
+   * original method public?
+   */
+  private List<Long> getFollowedSomething(long followerId, String fieldName) {
+    Filter userFilter = new FilterPredicate(FOLLOWER_FIELD_NAME, FilterOperator.EQUAL, followerId);
+    Filter otherFilter = new FilterPredicate(fieldName, FilterOperator.NOT_EQUAL, null);
+    Filter filter = CompositeFilterOperator.and(userFilter, otherFilter);
+    Query query = new Query("Follow").setFilter(filter);
+    PreparedQuery pq = datastore.prepare(query);
+
+    List<Long> list = new ArrayList<>();
+    for (Entity entity : pq.asIterable()) {
+      list.add((Long) entity.getProperty(fieldName));
+    }
+    return list;
+  }
+
   /** Retrieves deals posted by users followed by user */
   @Override
   public List<Deal> getDealsPublishedByFollowedUsers(long userId) {
-    return new ArrayList<Deal>();
+    return getDealsPublishedByFollowedRestaurantsOrUsers(userId, USER_FIELD_NAME, "posterId");
   }
 
   /** Retrieves deals posted by restaurants followed by user */
   @Override
   public List<Deal> getDealsPublishedByFollowedRestaurants(long userId) {
-    return new ArrayList<Deal>();
+    return getDealsPublishedByFollowedRestaurantsOrUsers(
+        userId, RESTAURANT_FIELD_NAME, "restaurantId");
   }
 
   /** Retrieves deals posted by tags followed by user */
   @Override
   public List<Deal> getDealsPublishedByFollowedTags(long userId) {
-    return new ArrayList<Deal>();
+    List<Deal> dealResults = new ArrayList<>();
+    List<Long> dealIdResults = new ArrayList<>();
+    List<Long> idsOfFollowedTags = followManager.getFollowedTagIds(userId);
+    for (Long id : idsOfFollowedTags) {
+      List<Long> dealIdsWithTag = dealTagManager.getDealIdsWithTag(id);
+      dealIdResults.addAll(dealIdsWithTag);
+    }
+    // Get rid of duplicate dealID (Deals with multiple tags)
+    List<Long> dealsWithoutDuplicates = new ArrayList<>(new HashSet<>(dealIdResults));
+    for (Long dealId : dealsWithoutDuplicates) {
+      dealResults.add(readDeal(dealId));
+    }
+    return dealResults;
   }
 
+  /** Sorts deals based on votes (Highest to lowest) */
   @Override
   public List<Deal> sortDealsBasedOnVotes(List<Deal> deals) {
-    return new ArrayList<Deal>();
+    Collections.sort(
+        deals,
+        new Comparator<Deal>() {
```

Comment:
the complexity of this is increased by calling `getVotes` every time we want to compare, because `getVotes` is an expensive operation. Instead, call `getVotes` once for each deal